src/Microsoft.VisualStudio.Threading/WeakKeyDictionary.cs

```diff
@@ -334,6 +361,106 @@ public override bool Equals(object obj)
             }
         }
 
+        /// <summary>
+        /// A helper structure to enumerate keys in the dictionary.
+        /// </summary>
+        internal class KeyEnumerable : IEnumerable<TKey>
+        {
+            private WeakKeyDictionary<TKey, TValue> dictionary;
+
+            internal KeyEnumerable(WeakKeyDictionary<TKey, TValue> dictionary)
+            {
+                Requires.NotNull(dictionary, nameof(dictionary));
+                this.dictionary = dictionary;
+            }
+
+            /// <summary>
+            /// Implements <see cref="IEnumerable{T}.GetEnumerator"/>
+            /// </summary>
+            IEnumerator<TKey> IEnumerable<TKey>.GetEnumerator()
+            {
+                return this.GetEnumerator();
+            }
+
+            /// <summary>
+            /// Implements <see cref="System.Collections.IEnumerable.GetEnumerator"/>
+            /// </summary>
+            System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
+            {
+                return this.GetEnumerator();
+            }
+
+            /// <summary>
+            /// Gets the Enumerator
+            /// </summary>
+            /// <returns>A new KeyEnumerator.</returns>
+            private KeyEnumerator GetEnumerator()
+            {
+                return new KeyEnumerator(this.dictionary);
+            }
+        }
+
+        /// <summary>
+        /// A helper structure to implement <see cref="IEnumerator{T}"/>
+        /// </summary>
+        private class KeyEnumerator : IEnumerator<TKey>
```

Comment:
Can we use a `struct` for this?

Reply:
I think all of our call path is going through the interface now, so a structure is likely causing more problem.